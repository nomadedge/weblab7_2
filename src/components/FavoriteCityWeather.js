import React, { Component } from 'react';
import { connect } from 'react-redux';

import { fetchCityWeather } from '../actions/FavoriteCitiesActions';

import Loader from './Loader';
import CityWeather from './CityWeather';

import '../styles/Weather.css';

class FavoriteCityWeather extends Component {
    componentDidMount() {
        const { cityName, fetchCityWeather } = this.props;
        fetchCityWeather(cityName);
    }

    render() {
        const { onDelete, favoriteCities } = this.props;
        const city = favoriteCities.find(city => city.name === this.props.cityName);

        if (city.isFetching) {
            return (
                <div id='favorite'>
                    <button id='delete-button' onClick={() => onDelete(city.name)}>Delete city</button>

                    <div id='weather-container'>
                        <div id='left-block'>
                            <div id='city-name'>{city.name}</div>
                        </div>
                        <div id='right-block'>
                            <Loader />
                        </div>
                    </div>
                </div>
            )
        }

        return (
            <>
                <div id='favorite'>
                    <button id='delete-button' onClick={() => onDelete(city.name)}>Delete city</button>

                    <div id='weather-container'>
                        <CityWeather cityWeather={city} />
                    </div>
                </div>
            </>
        );
    }
}

const mapStateToProps = ({ favoriteCities }) => {
    return {
        favoriteCities: favoriteCities.cities
    };
};
const mapDispatchToProps = {
    fetchCityWeather: fetchCityWeather
};

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteCityWeather);