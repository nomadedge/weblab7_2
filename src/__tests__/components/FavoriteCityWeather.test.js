import React from 'react';
import { Provider } from 'react-redux';
import FavoriteCityWeather from '../../components/FavoriteCityWeather';
import renderer from 'react-test-renderer';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

let existingArray;
beforeEach(() => {
    existingArray = [
        {
            name: 'correctCityName',
            isFetching: true,
            error: null,
            weather: {}
        }
    ];
});

describe('FavoriteCityWeather component', () => {
    test('matches the snapshot on loading', () => {
        const store = mockStore({
            favoriteCities: {
                cities: existingArray
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <FavoriteCityWeather cityName='correctCityName' />
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('matches the snapshot after loading for correct city', () => {
        existingArray[0].isFetching = false;
        existingArray[0].weather = {
            iconUrl: "http://openweathermap.org/img/w/50n.png",
            description: "mist",
            temperature: "14",
            pressure: 1022,
            humidity: 100,
            windSpeed: 3.6
        };

        const store = mockStore({
            favoriteCities: {
                cities: existingArray
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <FavoriteCityWeather cityName='correctCityName' />
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('matches the snapshot after loading error', () => {
        existingArray[0].isFetching = false;
        existingArray[0].error = 'Loading error.';

        const store = mockStore({
            favoriteCities: {
                cities: existingArray
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <FavoriteCityWeather cityName='correctCityName' />
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});