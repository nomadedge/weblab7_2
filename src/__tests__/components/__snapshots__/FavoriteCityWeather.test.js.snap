// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteCityWeather component matches the snapshot after loading error 1`] = `
<div
  id="favorite"
>
  <button
    id="delete-button"
    onClick={[Function]}
  >
    Delete city
  </button>
  <div
    id="weather-container"
  >
    <div
      id="left-block"
    >
      <div
        id="city-name"
      >
        correctCityName
      </div>
    </div>
    <div
      id="right-block"
    >
      <div>
        Loading error.
      </div>
    </div>
  </div>
</div>
`;

exports[`FavoriteCityWeather component matches the snapshot after loading for correct city 1`] = `
<div
  id="favorite"
>
  <button
    id="delete-button"
    onClick={[Function]}
  >
    Delete city
  </button>
  <div
    id="weather-container"
  >
    <div
      id="left-block"
    >
      <div
        id="city-name"
      >
        correctCityName
      </div>
      <div
        id="icon-temp"
      >
        <img
          alt="Weather icon"
          src="http://openweathermap.org/img/w/50n.png"
        />
        <div
          id="temp"
        >
          14
          Â°C
        </div>
      </div>
    </div>
    <div
      id="right-block"
    >
      <div
        className="info"
      >
        <div>
          Description:
        </div>
        <div
          className="value"
        >
          mist
        </div>
      </div>
      <div
        className="info"
      >
        <div>
          Barometer:
        </div>
        <div
          className="value"
        >
          1022
           hPa
        </div>
      </div>
      <div
        className="info"
      >
        <div>
          Humidity:
        </div>
        <div
          className="value"
        >
          100
           %
        </div>
      </div>
      <div
        className="info"
      >
        <div>
          Wind:
        </div>
        <div
          className="value"
        >
          3.6
           m/s
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FavoriteCityWeather component matches the snapshot on loading 1`] = `
<div
  id="favorite"
>
  <div
    id="weather-container"
  >
    <div
      id="left-block"
    >
      <div
        id="city-name"
      >
        correctCityName
      </div>
    </div>
    <div
      id="right-block"
    >
      <div
        id="spinner"
      />
    </div>
  </div>
</div>
`;
